### PR TITLE
Add TrackerAPI service and CreateMood endpoint

### PR DESCRIPTION
**Pull request by doniacld**
## Context

Create a simple mood tracker using Go and gRPC.
Empty endpoint for the moment.

## Changes

- Add proto definition of MoodAPI
- Add CreateMood endpoint to add a mood

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,30 @@
 # grpc-in-go
-Build microservices communicating thourgh gRPC in Golang
+
+Build a mood tracker using Go and gRPC.
+
+## Prerequisites
+
+Install [Go](https://go.dev/doc/install) 1.17
```
Comment (alienorlatour) [inline README.md]: `Latest stable version` will avoid the need to update this file every 6 months.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,30 @@
...
+
+Install protoc dependencies
```
Comment (alienorlatour) [inline README.md]: Tell me why, I want to know what I'm adding to my computer.
Also, can you have the CI generate the files ?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,30 @@
...
+    $ go install google.golang.org/protobuf/cmd/protoc-gen-go@latest
+    $ go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest
+
+Install [grpcurl](https://github.com/fullstorydev/grpcurl)
+
+    $ go install github.com/fullstorydev/grpcurl/cmd/grpcurl@latest
+
+## Run
+
+Run the project
```
Comment (alienorlatour) [inline README.md]:
```suggestion
Run the project. It will run on port 9000 by default.
```

File: cmd/api/protos/add-mood.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package protos
+
+import (
+	"context"
+	"log"
+
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type Server struct{}
```
Comment (alienorlatour) [inline cmd/api/protos/add-mood.go]: Exported means commented.

File: cmd/api/protos/add-mood.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+// AddMood adds a new mood.
```
Comment (alienorlatour) [inline cmd/api/protos/add-mood.go]: Does it add a Mood or a day's Log ?
Words are important.

File: cmd/api/protos/mood-api.proto [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+// MoodType, the possible values for a mood.
+enum MoodState {
+  Happy = 0;
+  Sad = 1;
+}
+
+// Mood definition.
+message Mood {
+  MoodState State = 1;
```
Comment (alienorlatour) [inline cmd/api/protos/mood-api.proto]: Needs doc

File: cmd/api/protos/mood-api.proto [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+package mood;
+
+import "google/protobuf/empty.proto";
+
+option go_package = "github.com/doniacld/grpc-in-go/cmd/api/protos";
+
+// MoodType, the possible values for a mood.
+enum MoodState {
+  Happy = 0;
```
Comment (alienorlatour) [inline cmd/api/protos/mood-api.proto]: Needs doc

File: cmd/api/protos/mood-api.proto [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  Sad = 1;
+}
+
+// Mood definition.
+message Mood {
+  MoodState State = 1;
+  // Date in epoch time.
+  int64 Date = 2;
```
Comment (alienorlatour) [inline cmd/api/protos/mood-api.proto]: I believe, from what I gather of your requirements, that the Date should be separate from the rest. Maybe a Mood object that will be saved for a given Date will make a Log. Then you can add Habits to the Log object. And then you can add PagesRead and Achievements and whatever.

File: internal/mood/app.go [hunk 9]
```diff
@@ -0,0 +1,4 @@
+package mood
+
+type App interface {
```
Comment (alienorlatour) [inline internal/mood/app.go]: What is this exactly? Do you need it?

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+GOGET=$(GOCMD) get
+GOMOD=$(GOCMD) mod
+GOGENERATE=$(GOCMD) generate
+GOLIST=$(GOCMD) list
+
+# Sources parameters
+SOURCE_ENTRYPOINT=./cmd/main.go
+
+# Binary parameters
+BINARY_NAME=main
```
Comment (floppyzedolfin) [inline Makefile]: Maybe give it a better name?

File: README.md [hunk 11]
```diff
@@ -1,2 +1,30 @@
...
+Build a mood tracker using Go and gRPC.
+
+## Prerequisites
+
+Install [Go](https://go.dev/doc/install) 1.17
+
+Install protoc dependencies
+
+    $ go install google.golang.org/protobuf/cmd/protoc-gen-go@latest
+    $ go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest
```
Comment (floppyzedolfin) [inline README.md:18]: The other option is to have a docker image that knows how to run protoc. But it requires some docker image storage for this.

File: README.md [hunk 12]
```diff
@@ -1,2 +1,30 @@
...
+
+## Run
+
+Run the project
+
+    $ make run
+
+### Example of request
+
+    > Getting the date in unix time : $ date +'%s'
```
Comment (floppyzedolfin) [inline README.md:34]: As a user, I think I'll be happier with `2022-01-30` in the API.

File: cmd/api/protos/add-mood.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package protos
```
Comment (floppyzedolfin) [inline cmd/api/protos/add-mood.go]: Dashes in file names ? I can't find the doc, but I remember something in the lines of "use underscores".
Or simply `addmood.go` ?

File: cmd/main.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	s := protos.Server{}
+
+	grpcServer := grpc.NewServer()
+
+	protos.RegisterMoodAPIServer(grpcServer, &s)
+
+	if err := grpcServer.Serve(lis); err != nil {
+		log.Fatalf("failed to serve: %s", err)
```
Comment (floppyzedolfin) [inline cmd/main.go]: Use `err.Error()` in the message

File: cmd/main.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+	"github.com/doniacld/grpc-in-go/cmd/api/protos"
+)
+
+func main() {
+	log.Println("Build a mood tracker using Go, gRPC and more!")
+
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", 9000))
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
```
Comment (floppyzedolfin) [inline cmd/main.go]: Use `err.Error()` when printing it.

File: go.mod [hunk 16]
```diff
@@ -0,0 +1,16 @@
+module github.com/doniacld/grpc-in-go
+
+go 1.18
```
Comment (floppyzedolfin) [inline go.mod:3]: But you said 1.17 :fearful:  !

File: README.md [hunk 17]
```diff
@@ -1,2 +1,30 @@
...
+
+    $ grpcurl --v -plaintext -import-path . \
+    --import-path proto/third_party -proto cmd/api/protos/mood-api.proto \
```
Comment (floppyzedolfin) [inline README.md]: Do you really need all these parameters?

File: README.md (hunk 3 above)
Comment (floppyzedolfin) [reply README.md]: Hit `Ctrl-C` to kill it.

File: README.md (hunk 11 above)
Comment (alienorlatour) [reply README.md:18]: How do people get into the official Docker registry?

Comment (floppyzedolfin) [reply README.md:18]: This is the tricky part. I don't know.
Also, this adds a  requirement on docker. 

Best leave it aside for now.

File: cmd/api/protos/mood-api.proto [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+// MoodAPI is the service holding mood operations methods.
+service MoodAPI {
+  // AddMood adds the given mood.
+  rpc AddMood (Mood) returns (google.protobuf.Empty);
```
Comment (floppyzedolfin) [inline cmd/api/protos/mood-api.proto]: We tend to go with `Create` : https://google.aip.dev/133

File: README.md (hunk 2 above)
Comment (doniacld) [reply README.md]: The goal is to make a tutorial to understand and learn at each step. 
Even though I can add it to the CI in a next step.

File: README.md (hunk 11 above)
Comment (doniacld) [reply README.md:18]: There is actually a registry in GitHub.

File: internal/mood/app.go (hunk 9 above)
Comment (doniacld) [reply internal/mood/app.go]: Not in this commit !

File: README.md (hunk 12 above)
Comment (doniacld) [reply README.md:34]: I am not sure if I make a front end I would be happy to call with a date as string or timestamp, it seems to be the easiest to have a timestamp. What do you think?

File: README.md (hunk 17 above)
Comment (doniacld) [reply README.md]: Nope ! Fixed.

File: go.mod (hunk 16 above)
Comment (doniacld) [reply go.mod:3]: I thought I was on Go 1.17 when developing.. 🧐

File: README.md [hunk 19]
```diff
@@ -1,30 +1,43 @@
...
+
+or install them manually like below.
+
+Install protoc dependencies to generate the client and the server in Go from the proto file
 
     $ go install google.golang.org/protobuf/cmd/protoc-gen-go@latest
     $ go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest
 
-Install [grpcurl](https://github.com/fullstorydev/grpcurl)
+Install [grpcurl](https://github.com/fullstorydev/grpcurl) to test easily gRPC request
```
Comment (alienorlatour) [inline README.md]: Prefer the adverb before the verb. To easily test.

File: README.md (hunk 12 above)
Comment (alienorlatour) [reply README.md:34]: Do you want, maybe, one day, to be able to add multiple entries in one day ?

File: cmd/api/protos/tracker-api.proto [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+  // Happy feeling
+  Happy = 0;
+  // Sad feeling
+  Sad = 1;
+}
+
+// Mood definition.
+message Mood {
+  // MoodState defines the mood state among defined values
+  MoodState State = 1;
```
Comment (alienorlatour) [inline cmd/api/protos/tracker-api.proto]: Should fields have a capital? What is the standard?
I usually use full lowercase to pass Goggle's horribly strong linter.

**Review by alienorlatour — COMMENTED** [review]
I much prefer this new data model.

File: cmd/api/protos/tracker-api.proto (hunk 20 above)
Comment (doniacld) [reply cmd/api/protos/tracker-api.proto]: Found the standard [gRPC standard language guide: Adding Comments](https://developers.google.com/protocol-buffers/docs/proto3#adding_comments)

**Pull request merged**